keps/sig-network/3836-kube-proxy-improved-ingress-connectivity-reliability/README.md

```diff
@@ -0,0 +1,346 @@
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The service controller in the Kubernetes cloud controller manager (KCCM)
+configures the load balancer and its corresponding health check (HC) following
+service related events. The configured health check is then used by the load
+balancer as to determine which instances are candidates for traffic load
+balancing. For certain cloud providers (GCP): the KCCM configures the HC to
+target the Kubernetes service proxy for this information. In this KEP we will
+focus on Kube-proxy, since it's the only service proxy under the responsibility
+of the Kubernetes project.
+
+We can define two classes of services for what concerns the HC:
+
+- A) services which have `externalTrafficPolicy: Cluster`
+- B) services which have `externalTrafficPolicy: Local`
+
+For class A) services: Kube-proxy currently returns an answer following its
+`healthz` state. For class B) services: Kube-proxy only reports if the service
+for which the load balancer was created for, has a `Ready` endpoint running on
+the node.
+
+This KEP proposes two changes:
+
+1. That Kube-proxy's includes the `healthz` state AND its current answer w.r.t
+  the local endpoints, when it answers to the HC for class B) services.
+
+2. That Kube-proxy enables connection draining for terminating Nodes. This is to
+  be done by Kube-proxy by inspecting a field on the Node object which indicates
+  that the "node is terminating/deleting" and when seen, starts failing the HC.
+
+```
+<<[UNRESOLVED HC-linked-service-fields ]>>
+We need to agree on this field. `deletionTimestamp` is the perfect indicator for
+this, but....the Node object does not have a finalizer, so that wouldn't help.
+@thockin advocates for `.spec.unschedulable`, but I am inclined to say "that
+doesn't indicate if a Node is going away". `unschedulable` just says that no
+workloads can be assigned to the Node anymore. It is true that it's usually
+followed up by a drain...but I am not sure it's good to implement something
+which can easily kill ingress connectivity to a Node / cluster by simply doing
+`kubectl cordon node ...` (It would kill ingress connectivity to the entire
+cluster if the ingress controller is running on the node which is being cordoned)
+
+Maybe we need a KEP for a finalizer on the Node object (*troll face)
```

Comment:
we can put our own finalizer, but it doesn't help in the case of CA.

Reply:
Yes, same mistake: diff between context in my mind and what I wrote. I meant  "improve Node EOL"

Reply:
it seems like we should at least file a feature request against sig-node...

Reply:
Already done - https://github.com/kubernetes/kubernetes/issues/115139